File: swift-backend/CodexDiscovery.swift
import Foundation

struct CodexSetup: Sendable {
    let path: String?
    let loggedIn: Bool
}

enum CodexDiscovery {
    static func status(checkLogin: Bool) -> CodexSetup {
        guard let path = resolve() else { return .init(path: nil, loggedIn: false) }
        return .init(path: path, loggedIn: checkLogin && run(path, arguments: ["login", "status"], timeout: 5))
    }

    static func resolve() -> String? {
        var candidates: [String] = []
        if let configured = ProcessInfo.processInfo.environment["CODEX_CLI_PATH"], !configured.isEmpty {
            candidates.append(configured)
        }
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        let path = ProcessInfo.processInfo.environment["PATH"] ?? ""
        candidates += path.split(separator: ":").map { "\($0)/codex" }
        candidates += [
            "\(home)/.local/bin/codex",
            "\(home)/.npm-global/bin/codex",
            "\(home)/.volta/bin/codex",
            "\(home)/.asdf/shims/codex",
            "\(home)/.local/share/mise/shims/codex",
            "/opt/homebrew/bin/codex",
            "/usr/local/bin/codex",
        ]
        let nvmRoot = "\(home)/.nvm/versions/node"
        if let versions = try? FileManager.default.contentsOfDirectory(atPath: nvmRoot) {
            candidates += versions.sorted(by: >).map { "\(nvmRoot)/\($0)/bin/codex" }
        }
        return candidates.first { FileManager.default.isExecutableFile(atPath: $0) }
    }

    private static func run(_ executable: String, arguments: [String], timeout: TimeInterval) -> Bool {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = arguments
        process.standardOutput = Pipe()
        process.standardError = Pipe()
        do { try process.run() } catch { return false }
        let semaphore = DispatchSemaphore(value: 0)
        process.terminationHandler = { _ in semaphore.signal() }
        if semaphore.wait(timeout: .now() + timeout) == .timedOut {
            process.terminate()
            return false
        }
        return process.terminationStatus == 0
    }
}

File: src/codex-discovery.js
import { accessSync, constants, readdirSync } from "node:fs";
import { homedir } from "node:os";
import { delimiter, dirname, join } from "node:path";

function executable(path) {
  if (!path) return false;
  try {
    accessSync(path, constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

function nvmCandidates(homeDirectory) {
  const root = join(homeDirectory, ".nvm", "versions", "node");
  try {
    return readdirSync(root, { withFileTypes: true })
      .filter((entry) => entry.isDirectory())
      .map((entry) => entry.name)
      .sort((left, right) => right.localeCompare(left, undefined, { numeric: true }))
      .map((version) => join(root, version, "bin", "codex"));
  } catch {
    return [];
  }
}

export function resolveCodexPath({
  environment = process.env,
  homeDirectory = homedir(),
  nodeExecutable = process.execPath,
} = {}) {
  const pathDirectories = String(environment.PATH ?? "")
    .split(delimiter)
    .map((path) => path.trim())
    .filter(Boolean);
  const candidates = [
    environment.CODEX_CLI_PATH,
    join(dirname(nodeExecutable), "codex"),
    ...pathDirectories.map((path) => join(path, "codex")),
    environment.npm_config_prefix ? join(environment.npm_config_prefix, "bin", "codex") : null,
    join(homeDirectory, ".local", "bin", "codex"),
    join(homeDirectory, ".npm-global", "bin", "codex"),
    join(homeDirectory, ".volta", "bin", "codex"),
    join(homeDirectory, ".bun", "bin", "codex"),
    join(homeDirectory, ".asdf", "shims", "codex"),
    join(homeDirectory, ".local", "share", "mise", "shims", "codex"),
    ...nvmCandidates(homeDirectory),
    "/opt/homebrew/bin/codex",
    "/usr/local/bin/codex",
    "/usr/bin/codex",
  ];
  for (const candidate of new Set(candidates.filter(Boolean))) {
    if (executable(candidate)) return candidate;
  }
  return null;
}
